crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -31,12 +31,39 @@ pub type BatchId = Blake3Digest<32>;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct TransactionBatch {
     id: BatchId,
-    updated_accounts: Vec<(TransactionId, TxAccountUpdate)>,
+    updated_accounts: BTreeMap<AccountId, AccountUpdate>,
     input_notes: Vec<InputNoteCommitment>,
     output_notes_smt: BatchNoteTree,
     output_notes: Vec<OutputNote>,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct AccountUpdate {
+    init_state: Digest,
+    final_state: Digest,
+    transactions: Vec<TransactionId>,
+    details: AccountUpdateDetails,
+}
+
+impl AccountUpdate {
+    fn new(tx: &ProvenTransaction) -> Self {
+        Self {
+            init_state: tx.account_update().init_state_hash(),
+            final_state: tx.account_update().final_state_hash(),
+            transactions: vec![tx.id()],
+            details: tx.account_update().details().clone(),
+        }
+    }
+
+    /// Merges the transaction's update into this account update.
+    fn merge_tx(&mut self, tx: &ProvenTransaction) {
+        self.final_state = tx.account_update().final_state_hash();
+        self.transactions.push(tx.id());
+
+        // TODO: merge account update details. Waiting on implementation in.. miden-objects.
```

Comment:
I need to merge `self.details` here but its not possible to do with the public API provided by `miden-objects` for `AccountUpdateDetails`.

Presumably such merge code would best live in that crate anyway since its fallible and involved.

Reply:
Could you make a PR with required changes in `miden-base`? I think the valid merge scenarios are:

1. `Private` + `Private` => `Private` (this should be trivial).
2. `Delta` + `Delta` => `Delta` (this is probably the most involved one since we'd need to merge 2 deltas).
3. `New` + `Delta` => `New` (this should be pretty easy because we can do `Account::apply_delta()`).

All other scenarios, as far as I can tell, should result in an error (i.e., `Delta` + `New` should not be possible as an account must first be created and then updated).

Reply:
Opened [#797](https://github.com/0xPolygonMiden/miden-base/pull/797)